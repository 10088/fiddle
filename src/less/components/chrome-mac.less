.darwin {
  .chrome {
    text-align: center;
    background-color: @background-1;
    border-bottom: @border;
    -webkit-app-region: drag;
    height: 23px;
    font-size: 1.4rem;

    small,
    p,
    span {
      user-select: none;
    }
  }
}
