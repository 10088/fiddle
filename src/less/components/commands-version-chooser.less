button.version-chooser.bp3-button {
  border-radius: 3px;
}
