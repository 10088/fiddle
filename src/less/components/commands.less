@import "../variables.less";

header {
  box-shadow: 0 0 0 1px rgba(16, 22, 26, 0.2), 0 0 0 rgba(16, 22, 26, 0), 0 1px 1px rgba(16, 22, 26, 0.4);
  background-color: @background-1;
  -webkit-app-region: drag;
}

.commands {
  display: flex;
  justify-content: space-between;
  height: 50px;
  padding: 10px;
  box-sizing: border-box;
  font-family: @fonts-common;

  button,
  input,
  select {
    -webkit-app-region: no-drag;
    margin-bottom: 0;
  }

  select {
    width: 100px;
    margin-left: 5px;
  }

  & > div {
    display: flex;

    &:last-of-type {
      margin-right: 10px;

      button {
        margin-left: 5px;
      }
    }
  }

  .address-bar {
    height: 28px;
    width: 360px;
    transition: width 0.4s;
    display: flex;

    input,
    span {
      font-family: @fonts-common;
      font-weight: 300;
      font-size: 11px;
      cursor: auto;
      -webkit-app-region: no-drag;
    }

    input {
      flex-grow: 1;
      background: white;
      padding-left: 1rem;
      height: 2.6rem;
    }

    input:invalid {
      box-shadow: inset 0 0 0 1px #e4252d;
    }

    span {
      background: white;
      color: #1d1d1d;
      border-bottom-color: rgb(209, 209, 209);
      border-bottom-style: solid;
      border-bottom-left-radius: 4px;
      border-bottom-width: 1px;
      border-image-outset: 0;
      border-image-repeat: stretch;
      border-image-slice: 100%;
      border-image-source: none;
      border-image-width: 1;
      border-left-color: rgb(209, 209, 209);
      border-left-style: solid;
      border-left-width: 1px;
      border-top-color: rgb(209, 209, 209);
      border-top-left-radius: 4px;
      border-top-style: solid;
      border-top-width: 1px;
      display: inline-block;
      height: 1.2rem;
      letter-spacing: normal;
      margin: 0;
      padding: 6px 0 10px 10px;
      user-select: text;
      width: 132px;
      text-rendering: auto;
      text-shadow: none;
      text-transform: none;
    }

    &.empty {
      width: 200px;
    }

    &:focus-within {
      span,
      input {
        border-color: @foreground-2;
      }
    }
  }
}
