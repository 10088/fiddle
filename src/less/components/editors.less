.mosaic-tile {
  animation: fadein 0.15s;
}

.mosaic-window-toolbar  {
  & > div {
    display: flex;
    justify-content: space-between;
    width: 100%;
    padding: 0 3px 0 29px;

    .mosaic-controls {
      transform: scale(0.7);
    }

    h5 {
      margin: 4px 0 0 0;
    }

    button {
      margin-left: 8px;
    }
  }
}

@keyframes fadein {
  from { opacity: 0; }
  to { opacity: 1; }
}
