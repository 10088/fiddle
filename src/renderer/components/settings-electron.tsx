import * as React from 'react';
import Icon from '@fortawesome/react-fontawesome';
import { faTrash, faCloudDownloadAlt, faSpinner } from '@fortawesome/fontawesome-free-solid';
import { observer } from 'mobx-react';

import { AppState } from '../state';
import { sortedElectronMap } from '../../utils/sorted-electron-map';
import { ElectronVersion } from '../../interfaces';
import { normalizeVersion } from '../../utils/normalize-version';

export interface ElectronSettingsProps {
  appState: AppState;
}

export interface ElectronSettingsState {
  isDownloadingAll: boolean;
  isDeletingAll: boolean;
}

/**
 * Settings content to manage Electron-related preferences.
 *
 * @class ElectronSettings
 * @extends {React.Component<ElectronSettingsProps, {}>}
 */
@observer
export class ElectronSettings extends React.Component<ElectronSettingsProps, ElectronSettingsState> {
  constructor(props: ElectronSettingsProps) {
    super(props);

    this.handleDownloadAll = this.handleDownloadAll.bind(this);
    this.handleDeleteAll = this.handleDeleteAll.bind(this);

    this.state = {
      isDownloadingAll: false,
      isDeletingAll: false
    };
  }

  /**
   * Download all versions of Electron.
   *
   * @returns {Promise<void>}
   */
  public async handleDownloadAll(): Promise<void> {
    this.setState({ isDownloadingAll: true });

    const { versions, downloadVersion } = this.props.appState;

    for (const key of Object.keys(versions)) {
      await downloadVersion(key);
    }

    this.setState({ isDownloadingAll: false });
  }

  /**
   * Delete all downloaded versions of Electron.
   *
   * @returns {Promise<void>}
   */
  public async handleDeleteAll(): Promise<void> {
    this.setState({ isDeletingAll: true });

    const { versions, removeVersion, version } = this.props.appState;

    for (const key in versions) {
      // If this isn't the currently selected version, remove it
      if (normalizeVersion(key) !== normalizeVersion(version)) {
        await removeVersion(key);
      }
    }

    this.setState({ isDeletingAll: false });
  }

  public render() {
    const { isDownloadingAll, isDeletingAll } = this.state;
    const isWorking = isDownloadingAll || isDeletingAll;
    const downloadIcon = isDownloadingAll
      ? <Icon icon={faSpinner} spin={true} />
      : <Icon icon={faCloudDownloadAlt} />;
    const deleteIcon = isDeletingAll
      ? <Icon icon={faSpinner} spin={true} />
      : <Icon icon={faCloudDownloadAlt} />;

    return (
      <div className='settings-electron'>
        <h2>Electron Settings</h2>
        <div className='advanced-options settings-section'>
          <label>Download all versions of Electron.</label>
          <button className='button' disabled={isWorking} onClick={this.handleDownloadAll}>
            {downloadIcon} Download All Versions
          </button>
          <label>Remove all downloaded versions of Electron.</label>
          <button className='button' disabled={isWorking} onClick={this.handleDeleteAll}>
            {deleteIcon} Delete All Versions
          </button>
        </div>
        <div>
          <h2>Versions</h2>
          {this.renderTable()}
        </div>
      </div>
    );
  }

  /**
   * Renders the table with Electron versions.
   *
   * @private
   * @returns {JSX.Element}
   */
  private renderTable(): JSX.Element {
    return (
      <table className='electron-versions-table'>
        <thead>
          <tr>
            <th>Version</th>
            <th>Status</th>
            <th className='action'>Action</th>
          </tr>
        </thead>
        <tbody>
          {this.renderTableRows()}
        </tbody>
      </table>
    );
  }

  /**
   * Renders the rows with Electron version, returning an Arary.
   *
   * @private
   * @returns {Array<JSX.Element>}
   */
  private renderTableRows(): Array<JSX.Element> {
    const { versions } = this.props.appState;

    return sortedElectronMap<JSX.Element>(versions, (key, item) => {
      const { tag_name, state } = item;
      const humanState = state === 'ready'
        ? '✅ Downloaded'
        : state === 'downloading' ? '⏬ Downloading' : '⏹ Not downloaded';

      return (
        <tr key={tag_name}>
          <td>{tag_name}</td>
          <td>{humanState}</td>
          <td>{this.renderAction(key, item)}</td>
        </tr>
      );
    });
  }

  /**
   * Renders the action for a single Electron version
   *
   * @private
   * @param {string} key
   * @param {ElectronVersion} item
   * @returns {JSX.Element}
   */
  private renderAction(key: string, item: ElectronVersion): JSX.Element {
    const { state } = item;
    const { appState } = this.props;

    // Already downloaded
    if (state === 'ready') {
      const remove = () => appState.removeVersion(key);

      return (
        <button className='button button-clear' onClick={remove}>
          <Icon icon={faTrash} /> Delete
        </button>
      );
    }

    // Downloading
    if (state === 'downloading') {
      return (
        <button className='button button-clear' disabled={true}>
          <Icon icon={faSpinner} spin={true}/> Downloading
        </button>
      );
    }

    // Unknown
    const download = () => appState.setVersion(key);

    return (
      <button className='button button-clear' onClick={download}>
        <Icon icon={faCloudDownloadAlt} /> Download
      </button>
    );
  }
}
