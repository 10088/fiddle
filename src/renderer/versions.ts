import { GitHubVersion } from '../interfaces';

/**
 * Retrieves our best guess regarding the latest Electron versions. Tries to
 * fetch them from localStorage, then from a static releases.json file.
 *
 * @returns {Array<GitHubVersion>}
 */
export function getKnownVersions(): Array<GitHubVersion> {
  const fromLs = window.localStorage.getItem('known-electron-versions');

  if (fromLs) {
    try {
      return JSON.parse(fromLs);
    } catch (error) {
      console.warn(`Parsing known versions failed, falling back`);
    }
  }

  return require('../../static/releases.json');
}

/**
 * Saves known versions to localStorage.
 *
 * @param {Array<GitHubVersion>} versions
 */
export function saveKnownVersions(versions: Array<GitHubVersion>) {
  const stringified = JSON.stringify(versions);
  window.localStorage.setItem('known-electron-versions', stringified);
}

/**
 * Tries to refresh our known versions and returns whatever we have
 * saved after.
 *
 * @export
 * @returns {Promise<Array<GitHubVersion>>}
 */
export async function getUpdatedKnownVersions(
  pages: number
): Promise<Array<GitHubVersion>> {
  try {
    await fetchVersions(pages);
  } catch (error) {
    console.warn(`Versions: Failed to fetch versions`, { error });
  }

  return getKnownVersions();
}

/**
 * Fetch the latest known versions directly from GitHub.
 *
 * @returns {Promise<Array<GitHubVersion>>}
 */
export async function fetchVersions(pages: number) {
  const output: Array<GitHubVersion> = [];

  for (let i = 0; i < pages; i++) {
    const page = await fetchVersionPage(i + 1);

    if (page && page.length > 0 && page[0].tag_name) {
      page.forEach((release) => {
        const updated = {
          url: release.url,
          assets_url: release.assets_url,
          html_url: release.html_url,
          tag_name: release.tag_name,
          target_commitish: release.target_commitish,
          name: release.name,
          prerelease: release.prerelease,
          created_at: release.created_at,
          published_at: release.published_at,
          body: release.body
        };

        output.push(updated as GitHubVersion);
      });
    }
  }

  saveKnownVersions(output);
}

/**
 * Fetch a single releases page from GitHub.
 *
 * @returns {Promise<Array<GitHubVersion>>}
 */
export function fetchVersionPage(page?: number): Promise<Array<GitHubVersion>> {
  const url = `https://api.github.com/repos/electron/electron/releases`
    + ((page && page > 1) ? `?page=${page}` : '');

  return window.fetch(url)
    .then((response) => response.json());
}
