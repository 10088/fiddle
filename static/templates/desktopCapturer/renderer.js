const { desktopCapturer } = require('electron')

// The following example shows how to capture video from
// the screen. It also grabs each window, so you could
// just grab video from a single window.
//
desktopCapturer.getSources({ types: ['window', 'screen'] }, (error, sources) => {
  if (error) throw error
  for (let i = 0; i < sources.length; ++i) {
    console.log(sources[i])

    if (sources[i].name.startsWith('Screen')) {
      navigator.mediaDevices.getUserMedia({
        audio: false,
        video: {
          mandatory: {
            chromeMediaSource: 'desktop',
            chromeMediaSourceId: sources[i].id,
            minWidth: 1280,
            maxWidth: 1280,
            minHeight: 720,
            maxHeight: 720
          }
        }
      })
      .then((stream) => handleStream(stream))
      .catch((e) => handleError(e))
      return
    }
  }
})

function handleStream (stream) {
  const video = document.querySelector('video')
  video.srcObject = stream
  video.onloadedmetadata = (e) => video.play()
}

function handleError (e) {
  console.log(e)
}