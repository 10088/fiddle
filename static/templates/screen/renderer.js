const { screen } = require('electron')

const screens = screen.getAllDisplays()
const div = document.querySelector('div')

div.innerText = JSON.stringify(screens);
