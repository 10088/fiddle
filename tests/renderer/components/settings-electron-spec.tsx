import { shallow } from 'enzyme';
import * as React from 'react';

import { ElectronSettings } from '../../../src/renderer/components/settings-electron';
import { ElectronReleaseChannel } from '../../../src/renderer/versions';
import { mockVersions } from '../../mocks/electron-versions';

describe('ElectronSettings component', () => {
  let store: any;

  beforeEach(() => {
    store = {
      version: '2.0.1',
      versions: mockVersions,
      versionPagesToFetch: 2,
      versionsToShow: [ ElectronReleaseChannel.stable, ElectronReleaseChannel.beta ],
      downloadVersion: jest.fn(),
      removeVersion: jest.fn(),
      updateElectronVersions: jest.fn(),
      toggleAddVersionDialog: jest.fn()
    };

    // Render all the states
    store.versions['2.0.2'].state = 'downloading';
    store.versions['2.0.1'].state = 'ready';
    store.versions['1.8.7'].state = 'unknown';
  });

  it('renders', () => {
    const wrapper = shallow(
      <ElectronSettings appState={store} />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('handles the deleteAll()', async () => {
    const wrapper = shallow(
      <ElectronSettings appState={store} />
    );
    const instance = wrapper.instance() as any;
    await instance.handleDeleteAll();

    expect(store.removeVersion).toHaveBeenCalledTimes(2);
  });

  it('handles the downloadAll()', async () => {
    const wrapper = shallow(
      <ElectronSettings appState={store} />
    );
    const instance = wrapper.instance() as any;
    await instance.handleDownloadAll();

    expect(store.downloadVersion).toHaveBeenCalled();
  });

  describe('handleDownloadClick()', () => {
    it('kicks off an update of Electron versions', async () => {
      const wrapper = shallow(
        <ElectronSettings appState={store} />
      );
      const instance = wrapper.instance() as any;
      await instance.handleDownloadClick();

      expect(store.updateElectronVersions).toHaveBeenCalled();
    });
  });

  describe('handleAddVersion()', () => {
    it('toggles the add version dialog', () => {
      const wrapper = shallow(
        <ElectronSettings appState={store} />
      );
      const instance = wrapper.instance() as any;
      instance.handleAddVersion();

      expect(store.toggleAddVersionDialog).toHaveBeenCalled();
    });
  });

  describe('handlePagesChange()', () => {
    it('handles a new selection', async () => {
      const wrapper = shallow(
        <ElectronSettings appState={store} />
      );
      const instance = wrapper.instance() as any;
      await instance.handlePagesChange({
        currentTarget: {
          value: '120'
        }
      });

      expect(store.versionPagesToFetch).toBe(4);
    });
  });

  describe('handleChannelChange()', () => {
    it('handles a new selection', async () => {
      const wrapper = shallow(
        <ElectronSettings appState={store} />
      );
      const instance = wrapper.instance() as any;
      await instance.handleChannelChange({
        currentTarget: {
          id: ElectronReleaseChannel.stable,
          checked: false
        }
      });

      await instance.handleChannelChange({
        currentTarget: {
          id: ElectronReleaseChannel.nightly,
          checked: true
        }
      });

      expect(store.versionsToShow).toEqual([
        ElectronReleaseChannel.beta,
        ElectronReleaseChannel.nightly
      ]);
    });
  });
});
