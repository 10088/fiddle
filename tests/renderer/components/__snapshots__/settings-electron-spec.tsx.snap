// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ElectronSettings component renders 1`] = `
<div
  className="settings-electron"
>
  <h2>
    Electron Settings
  </h2>
  <Blueprint3.Callout>
    <Blueprint3.FormGroup
      label="Include Electron versions from these release channels:"
    >
      <Blueprint3.Checkbox
        checked={true}
        id="Stable"
        inline={true}
        label="Stable"
        onChange={[Function]}
      />
      <Blueprint3.Checkbox
        checked={true}
        id="Beta"
        inline={true}
        label="Beta"
        onChange={[Function]}
      />
      <Blueprint3.Checkbox
        checked={false}
        id="Nightly"
        inline={true}
        label="Nightly"
        onChange={[Function]}
      />
      <Blueprint3.Checkbox
        checked={false}
        id="Unsupported"
        inline={true}
        label="Unsupported"
        onChange={[Function]}
      />
    </Blueprint3.FormGroup>
    <Blueprint3.FormGroup
      label="Number of recent Electron releases to include:"
    >
      <Blueprint3.Slider
        disabled={false}
        initialValue={30}
        labelStepSize={30}
        max={300}
        min={30}
        onChange={[Function]}
        showTrackFill={true}
        stepSize={30}
        value={60}
        vertical={false}
      />
    </Blueprint3.FormGroup>
  </Blueprint3.Callout>
  <br />
  <Blueprint3.Callout>
    <Blueprint3.ButtonGroup
      fill={true}
    >
      <Blueprint3.Button
        icon="numbered-list"
        onClick={[Function]}
        text="Update Electron Release List"
      />
      <Blueprint3.Button
        disabled={false}
        icon="download"
        onClick={[Function]}
        text="Download All Versions"
      />
      <Blueprint3.Button
        disabled={false}
        icon="trash"
        onClick={[Function]}
        text="Delete All Downloads"
      />
      <Blueprint3.Button
        icon="document-open"
        onClick={[Function]}
        text="Add Local Electron Build"
      />
    </Blueprint3.ButtonGroup>
    <HTMLTable
      className="electron-versions-table"
      striped={true}
    >
      <thead>
        <tr>
          <th>
            Version
          </th>
          <th>
            Status
          </th>
          <th
            className="action"
          >
            Action
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          key="v2.0.2"
        >
          <td>
            v2.0.2
          </td>
          <td>
            <span>
              <Blueprint3.Icon
                icon="cloud-download"
              />
               
              Downloading
            </span>
          </td>
          <td
            className="action"
          >
            <Blueprint3.Button
              disabled={true}
              fill={true}
              icon="cloud-download"
              loading={true}
              small={true}
              text="Downloading"
            />
          </td>
        </tr>
        <tr
          key="v2.0.1"
        >
          <td>
            v2.0.1
          </td>
          <td>
            <span>
              <Blueprint3.Icon
                icon="box"
              />
               
              Downloaded
            </span>
          </td>
          <td
            className="action"
          >
            <Blueprint3.Button
              fill={true}
              icon="trash"
              onClick={[Function]}
              small={true}
              text="Delete"
            />
          </td>
        </tr>
        <tr
          key="v1.8.7"
        >
          <td>
            v1.8.7
          </td>
          <td>
            <span>
              <Blueprint3.Icon
                icon="cloud"
              />
               
              Not downloaded
            </span>
          </td>
          <td
            className="action"
          >
            <Blueprint3.Button
              disabled={false}
              fill={true}
              icon="cloud-download"
              loading={false}
              onClick={[Function]}
              small={true}
              text="Download"
            />
          </td>
        </tr>
      </tbody>
    </HTMLTable>
  </Blueprint3.Callout>
</div>
`;
